File: foliage-rpc/src/main/java/com/github/lavenderx/netty_rpc/registry/ServiceDiscovery.java
package com.github.lavenderx.netty_rpc.registry;

import com.github.lavenderx.netty_rpc.client.ConnectionManager;
import lombok.extern.slf4j.Slf4j;
import org.apache.zookeeper.KeeperException;
import org.apache.zookeeper.Watcher;
import org.apache.zookeeper.ZooKeeper;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ThreadLocalRandom;

@Slf4j
public class ServiceDiscovery {

    private final CountDownLatch latch = new CountDownLatch(1);
    private final String registryAddress;
    private final ZooKeeper zookeeper;

    private volatile List<String> dataList = new ArrayList<>();

    public ServiceDiscovery(String registryAddress) {
        this.registryAddress = registryAddress;
        this.zookeeper = connectServer();
        if (this.zookeeper != null) {
            watchNode(this.zookeeper);
        }
    }

    public String discover() {
        String data = null;
        int size = dataList.size();
        if (size > 0) {
            if (size == 1) {
                data = dataList.get(0);
                log.debug("using only data: {}", data);
            } else {
                data = dataList.get(ThreadLocalRandom.current().nextInt(size));
                log.debug("using random data: {}", data);
            }
        }
        return data;
    }

    private ZooKeeper connectServer() {
        ZooKeeper zk = null;
        try {
            zk = new ZooKeeper(registryAddress, Constants.ZK_SESSION_TIMEOUT, event -> {
                if (event.getState() == Watcher.Event.KeeperState.SyncConnected) {
                    latch.countDown();
                }
            });
            latch.await();
        } catch (IOException | InterruptedException ex) {
            log.error("", ex);
        }
        return zk;
    }

    private void watchNode(final ZooKeeper zk) {
        try {
            List<String> nodeList = zk.getChildren(Constants.ZK_REGISTRY_PATH, event -> {
                if (event.getType() == Watcher.Event.EventType.NodeChildrenChanged) {
                    watchNode(zk);
                }
            });
            List<String> dataList = new ArrayList<>();
            for (String node : nodeList) {
                byte[] bytes = zk.getData(Constants.ZK_REGISTRY_PATH + "/" + node, false, null);
                dataList.add(new String(bytes));
            }
            log.debug("node data: {}", dataList);
            this.dataList = dataList;

            log.debug("Service discovery triggered updating connected server node.");
            UpdateConnectedServer();
        } catch (KeeperException | InterruptedException ex) {
            log.error("", ex);
        }
    }

    private void UpdateConnectedServer() {
        ConnectionManager.getInstance().updateConnectedServer(this.dataList);
    }

    public void stop() {
        if (this.zookeeper != null) {
            try {
                this.zookeeper.close();
            } catch (InterruptedException ex) {
                log.error("", ex);
            }
        }
    }
}

File: foliage-rpc/src/main/kotlin/org/lavenderx/foliage/nettyrpc/registry/ServiceDiscovery.kt
package org.lavenderx.foliage.nettyrpc.registry

import org.apache.zookeeper.KeeperException
import org.apache.zookeeper.ZooKeeper
import org.lavenderx.foliage.nettyrpc.client.ConnectionManager
import org.lavenderx.foliage.nettyrpc.logging.loggerFor
import java.util.*
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ThreadLocalRandom

class ServiceDiscovery(private val registryAddress: String) {

    private val logger = loggerFor<ServiceDiscovery>()
    private val latch = CountDownLatch(1)
    private val zookeeper: ZooKeeper?
    @Volatile private var dataList: List<String> = ArrayList()

    init {
        this.zookeeper = connectServer()
        if (this.zookeeper != null) {
            watchNode(this.zookeeper)
        }
    }


    fun discover(): String {
        var data: String? = null
        val size = dataList.size
        if (size > 0) {
            if (size == 1) {
                data = dataList[0]
                logger.debug("using only data: {}", data)
            } else {
                data = dataList[ThreadLocalRandom.current().nextInt(size)]
                logger.debug("using random data: {}", data)
            }
        }
        return data!!
    }

    private fun connectServer(): ZooKeeper? {
        var zk: ZooKeeper? = null
        try {
            zk = ZooKeeper(registryAddress, Constants.ZK_SESSION_TIMEOUT) { event ->
                if (event.state == org.apache.zookeeper.Watcher.Event.KeeperState.SyncConnected) {
                    latch.countDown()
                }
            }
            latch.await()
        } catch (e: Exception) {
            logger.error("", e)
        }

        return zk
    }

    private fun watchNode(zk: ZooKeeper) {
        try {
            val nodeList = zk.getChildren(Constants.ZK_REGISTRY_PATH) { event ->
                if (event.type == org.apache.zookeeper.Watcher.Event.EventType.NodeChildrenChanged) {
                    watchNode(zk)
                }
            }
            val dataList = ArrayList<String>()
            for (node in nodeList) {
                val bytes = zk.getData(Constants.ZK_REGISTRY_PATH + "/" + node, false, null)
                dataList.add(String(bytes))
            }
            logger.debug("node data: {}", dataList)
            this.dataList = dataList

            logger.debug("Service discovery triggered updating connected server node.")
            UpdateConnectedServer()
        } catch (e: KeeperException) {
            logger.error("", e)
        }

    }

    private fun UpdateConnectedServer() {
        ConnectionManager.getInstance().updateConnectedServer(this.dataList)
    }

    fun stop() {
        try {
            this.zookeeper?.close()
        } catch (e: InterruptedException) {
            logger.error("", e)
        }
    }
}